### PR TITLE
Format metrics; add (min, med, max) for timing metrics

### PR DESCRIPTION
**Pull request by vishalkarve15**
Shows up like this now: 
Screenshot 2023-08-29 at 9 40 49 PM

### CODE REVIEW

File: spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryParseTimeMetric.java [hunk 1]
```diff
@@ -15,4 +17,9 @@ public String name() {
   public String description() {
     return BIG_QUERY_PARSE_TIME_METRIC_DESCRIPTION;
   }
+
+  @Override
+  public String aggregateTaskMetrics(long[] taskMetrics) {
+    return stringValue(taskMetrics, MetricsType.TIMING_METRIC);
```
Comment (arunkumarchacko) [inline spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryParseTimeMetric.java:23]: nit: I do not see the value of having the MetricsType enum. Instead, consider having separate methods for each of the enum values.

Comment (davidrabinowitz) [reply spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryParseTimeMetric.java:23]: +1

File: spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryCustomMetricConstants.java [hunk 2]
```diff
@@ -6,10 +6,9 @@ public class SparkBigQueryCustomMetricConstants {
   public static final String BIG_QUERY_ROWS_READ_METRIC_NAME = "bqRowsRead";
   static final String BIG_QUERY_ROWS_READ_METRIC_DESCRIPTION = "number of BQ rows read";
   public static final String BIG_QUERY_SCAN_TIME_METRIC_NAME = "bqScanTime";
-  static final String BIG_QUERY_SCAN_TIME_METRIC_DESCRIPTION = "scan time for BQ in milli sec";
+  static final String BIG_QUERY_SCAN_TIME_METRIC_DESCRIPTION = "scan time for BQ";
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryCustomMetricConstants.java:9]: It's in the UI only, makes it less cluttered

File: spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/MetricUtils.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import java.text.NumberFormat;
+import java.util.Arrays;
+import java.util.Locale;
+
+public class MetricUtils {
+
+  static final String METRIC_FORMAT = "total (min, med, max)\n%s (%s, %s, %s)";
+
+  private static String toNumberFormat(long value) {
+    NumberFormat numberFormat = NumberFormat.getIntegerInstance(Locale.US);
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/MetricUtils.java:15]: +1

File: spark-bigquery-dsv2/spark-bigquery-metrics/src/test/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryScanTimeMetricTest.java [hunk 4]
```diff
@@ -22,7 +23,7 @@ public void testDescription() {
 
   @Test
   public void testAggregateMetrics() {
-    assertThat(sparkBigQueryScanTimeMetric.aggregateTaskMetrics(new long[] {1000L, 2000L}))
-        .isEqualTo("3000");
+    assertThat(sparkBigQueryScanTimeMetric.aggregateTaskMetrics(new long[] {1010L, 2020L, 3030L}))
+        .isEqualTo(String.format(METRIC_FORMAT, "6.1 s", "1.0 s", "2.0 s", "3.0 s"));
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-bigquery-metrics/src/test/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryScanTimeMetricTest.java:27]: Please replace the String.format() with the actual String

File: spark-bigquery-dsv2/spark-bigquery-metrics/src/test/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryTimeInSparkMetricTest.java [hunk 5]
```diff
@@ -23,7 +24,8 @@ public void testDescription() {
 
   @Test
   public void testAggregateMetrics() {
-    assertThat(sparkBigQueryTimeInSparkMetric.aggregateTaskMetrics(new long[] {1000L, 2000L}))
-        .isEqualTo("3000");
+    assertThat(
+            sparkBigQueryTimeInSparkMetric.aggregateTaskMetrics(new long[] {1020L, 2010L, 3030L}))
+        .isEqualTo(String.format(METRIC_FORMAT, "6.1 s", "1.0 s", "2.0 s", "3.0 s"));
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-bigquery-metrics/src/test/java/com/google/cloud/spark/bigquery/v2/customMetrics/SparkBigQueryTimeInSparkMetricTest.java:29]: Please replace the String.format() with the actual String

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/MetricUtils.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    Arrays.sort(taskMetrics);
+    String sum = msDurationToString(Arrays.stream(taskMetrics).sum());
+    String min = msDurationToString(taskMetrics[0]);
+    String med = msDurationToString(taskMetrics[taskMetrics.length / 2]);
+    String max = msDurationToString(taskMetrics[taskMetrics.length - 1]);
+    return String.format(METRIC_FORMAT, sum, min, med, max);
+  }
+
+  static String stringValue(long[] taskMetrics, MetricsType metricsType) {
```
[the remark this answers is not in the document]
Comment (vishalkarve15) [reply spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/MetricUtils.java:34]: moving them to separate methods.

File: spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/MetricsType.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package com.google.cloud.spark.bigquery.v2.customMetrics;
+
+public enum MetricsType {
```
[the remark this answers is not in the document]
Comment (vishalkarve15) [reply spark-bigquery-dsv2/spark-bigquery-metrics/src/main/java/com/google/cloud/spark/bigquery/v2/customMetrics/MetricsType.java:3]: removing this.

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**